cmd/snap/cmd_warnings.go

```diff
@@ -0,0 +1,210 @@
+	Verbose bool `long:"verbose"`
+}
+
+type cmdOkay struct{ clientMixin }
+
+var shortWarningsHelp = i18n.G("List warnings")
+var longWarningsHelp = i18n.G(`
+The warnings command lists the warnings that have been reported to the system.
+
+Once warnings have been listed with 'snap warnings', 'snap okay' may be used to
+silence them. A warning that's been silenced in this way will not be listed
+again unless it happens again, _and_ a cooldown time has passed.
+
+Warnings expire automatically, and once expired they are forgotten.
+`)
+
+var shortOkayHelp = i18n.G("Acknowledge warnings")
+var longOkayHelp = i18n.G(`
+The okay command acknowledges the warnings listed with 'snap warnings'.
+
+Once acknowledged a warning won't appear again unless it re-occurrs and
+sufficient time has passed.
+`)
+
+func init() {
+	addCommand("warnings", shortWarningsHelp, longWarningsHelp, func() flags.Commander { return &cmdWarnings{} }, timeDescs.also(map[string]string{
+		"all":     i18n.G("Show all warnings"),
+		"verbose": i18n.G("Show more information"),
+	}), nil)
+	addCommand("okay", shortOkayHelp, longOkayHelp, func() flags.Commander { return &cmdOkay{} }, nil, nil)
+}
+
+func (cmd *cmdWarnings) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	warnings, err := cmd.client.Warnings(client.WarningsOptions{All: cmd.All})
+	if err != nil {
+		return err
+	}
+	if len(warnings) == 0 {
+		fmt.Fprintln(Stderr, i18n.G("No new warnings."))
+		return nil
+	}
+
+	if err := writeWarningTimestamp(time.Now()); err != nil {
+		return err
+	}
+
+	w := tabWriter()
+	if cmd.Verbose {
+		fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%s\n",
+			i18n.G("First occurrence"),
+			i18n.G("Last occurrence"),
+			i18n.G("Expires after"),
+			i18n.G("Acknowledged"),
+			i18n.G("Repeats after"),
+			i18n.G("Warning"))
+		for _, warning := range warnings {
+			lastShown := "-"
+			if warning.LastShown != nil && !warning.LastShown.IsZero() {
+				lastShown = cmd.fmtTime(*warning.LastShown)
+			}
+			fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%s\n",
+				cmd.fmtTime(warning.FirstAdded),
+				cmd.fmtTime(warning.LastAdded),
+				warning.ExpireAfter,
+				lastShown,
+				warning.RepeatAfter,
+				warning.Message)
+		}
+	} else {
+		fmt.Fprintf(w, "%s\t%s\n", i18n.G("Last occurrence"), i18n.G("Warning"))
+		for _, warning := range warnings {
+			fmt.Fprintf(w, "%s\t%s\n", cmd.fmtTime(warning.LastAdded), warning.Message)
+		}
+	}
+	w.Flush()
+
+	return nil
+}
+
+func (cmd *cmdOkay) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	last, err := lastWarningTimestamp()
+	if err != nil {
+		return fmt.Errorf("no client-side warning timestamp found: %v", err)
+	}
+
+	return cmd.client.Okay(last)
+}
+
+const warnFileEnvKey = "SNAPD_LAST_WARNING_TIMESTAMP_FILENAME"
+
+func warnFilename(homedir string) string {
+	if fn := os.Getenv(warnFileEnvKey); fn != "" {
+		return fn
+	}
+
+	return filepath.Join(homedir, ".snap", "warning")
```

Comment:
(nitpick) Should we use plural `warnings` (warning**s**) here?

Reply:
Nitpick #2: we already keep `auth.json` in this directory, would it make sense to use .json extension here as well for consistency?